hw/scsi/scsi-disk.c

```diff
@@ -570,12 +573,68 @@ static uint8_t *scsi_get_buf(SCSIRequest *req)
     return (uint8_t *)r->iov.iov_base;
 }
 
+typedef struct __page
+{
+    uint8_t page_code;
+    uint8_t sub_page_code;
+    uint16_t offset;
+    uint16_t length;
+}PageIndex;
+
+typedef struct __pages
+{
+    uint16_t number;
+    PageIndex pi;
+}PagesHeader;
+
+typedef struct __file
+{
+    uint16_t inquiry_offset;
+    uint16_t mode_offset;
+}FileHeader;
+
+static int page_load(SCSIDiskState *s, int page, int sub_page, uint8_t **p_outbuf,
+                                int page_type, int *available_space)
+{
+
+    int length = -1;
+    FileHeader* header = (FileHeader*)s->page_buffer;
+    PagesHeader* p_pg_header = (PagesHeader*)(s->page_buffer + *(&(header->inquiry_offset) + page_type));  // get ptr to Pages Header.
+    PageIndex* index = &p_pg_header->pi;                                             // get the index to first page.
+    for (int i = 0; i< p_pg_header->number; i++){
+        if (page == index->page_code && sub_page == index->sub_page_code) {             // find the correct page and sub page.
+            if (index->length > *available_space) {
+                break;
+            }
+            memcpy(*p_outbuf, (uint8_t*)p_pg_header + index->offset, index->length);
+            length = index->length;
+            *available_space -= length;
+            *p_outbuf += length;
+            break;
+        }
+        ++index;
+    }
+    return length;
+}
+
 static int scsi_disk_emulate_inquiry(SCSIRequest *req, uint8_t *outbuf)
 {
     SCSIDiskState *s = DO_UPCAST(SCSIDiskState, qdev, req->dev);
+    SCSIDiskReq *r = DO_UPCAST(SCSIDiskReq, req, req);
+
     int buflen = 0;
     int start;
-
+    if (s->page_buffer) {
+        int available_space = r->buflen;
+        if (req->cmd.buf[1] & 0x1) {
+            // VPD Page
+            buflen = page_load(s, req->cmd.buf[2], 0, &outbuf, 0, &available_space);
+        } else {
+            // Standard Page
+            buflen = page_load(s, 0xff, 0xff, &outbuf, 0, &available_space);
```

Comment:
No, because there is no page code for standard page, and we know that INQUIRY Pages have no subpages. This is the reason we use page code = 0xff, sub page code = 0xff for Standard Inquiry page, so that we could put it into "inquiry" group in our generated bin file without create another space.